### PR TITLE
Update Settings: Update VPN settings item

### PR DESCRIPTION
**Pull request by mikescamell**
Task/Issue URL: https://app.asana.com/0/0/1208966262488271/f 

### Description

Updates the VPN settings item to match new designs. As the new designs require us to know more about the subscription state we need to expose more Subscription Statuses so we can reflect that in the settings screen like the "Waiting" state that is equivalent to our new "Activating" state.

The biggest thing to note is that now it's not like it used to be where we had just 2 states, locked and unlocked, we now more readily show the user the VPN item and what it's subscription state is.

I'm still waiting on answers regarding the [onboarded state](https://app.asana.com/0/0/1208847942321390/1208973040958179/f)

And [connecting state ](https://app.asana.com/0/0/1208847942321390/1208973157895050/f) but I can always circle back later and address these as necessary.

[Designs](https://www.figma.com/design/CjH849hL53lhsPlf6Ufeo4/%E2%9A%99%EF%B8%8F-Browser-Settings-Documentation-(All-Platforms)?node-id=7605-431390&node-type=instance&t=ZKIMXkdZqzh5gEjU-11)

### Steps to test this PR

Prerequisite: Enable `newSettings` feature toggle

Prerequisite: Apply the patch in the [Asana task](https://app.asana.com/0/1207908166761516/1208966262488271/f)

Note: Only the VPN item and possibly the Settings item will be seen, depending on the different states. PIR, ITR are not included as part of this PR and so are not visible.

_VPN Subscribed and Off_
- [x] Open New Settings
- [x] Verify VPN item is visible with colored icon and status is off
- [x] Click VPN item
- [x] VPN settings screen should open

_VPN Subscribed and On_
- [x] In `NetworkProtectionStateImpl` change ln 98 to `return flowOf(CONNECTED)`
- [x] Open New Settings
- [x] Verify VPN item is visible with colored icon and status is on
- [x] Click VPN item
- [x] VPN settings screen should open

_Unsubscribed_
- [x] In `SubscriptionManager` change ln 442 to `return SubscriptionStatus.UNKNOWN` 
- [x] Open New Settings
- [x] Verify VPN item is not visible

_VPN Expired_
- [x] In `SubscriptionManager` change ln 442 to `return SubscriptionStatus.EXPIRED` 
- [x] Open New Settings
- [x] Verify VPN item is visible with grey icon and is  not clickable, and status is off
- [x] Click VPN item
- [x] Nothing should happen

_PPro Activating_
- [x] In `SubscriptionManager` change ln 442 to `return SubscriptionStatus.WAITING` 
- [x] Open New Settings
- [x] Verify VPN item is visible with grey icon and status is off
- [x] Click VPN item
- [x] Nothing should happen

_No Entitlement_
- [x] In `SubscriptionManager` change ln 442 to `return SubscriptionStatus.AUTO_RENEWABLE` 
- [x] In `RealSubscriptions` change ln 77 to `return flowOf(listOf())` 
- [x] Open New Settings
- [x] Verify VPN item is not visible

_Legacy Support_
- [x] Turn off `newSettings`
- [x] Verify old VPN settings works by repeating steps above

### UI changes
| Before  | After |
| --- | --- |
| old_vpn_active  | vpn_subscribed_on |
| old_vpn_disconnected | vpn_subscribed_off |
| old_vpn_waiting | vpn_activating |
| old_vpn_ineligible | new_vpn_ineligible | old_no_subscription | vpn_no_subscription |

### CODE REVIEW

**Comment by mikescamell** [comment]
* **#5394** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5394?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/5394?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `develop`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: network-protection/network-protection-api/src/main/java/com/duckduckgo/networkprotection/api/NetworkProtectionAccessState.kt [hunk 1]
```diff
@@ -25,13 +25,25 @@ interface NetworkProtectionAccessState {
...
      * The caller DOES NOT need to specify the dispatcher when calling this method
      */
-    suspend fun getStateFlow(): Flow<NetPAccessState>
+    suspend fun getLegacyStateFlow(): Flow<LegacyNetPAccessState>
+
+    /**
+     * Returns the visibility state of the NetP
+     * The caller DOES NOT need to specify the dispatcher when calling this method
+     */
+    suspend fun getState(): NetPVisibilityState
```
Comment (malmstein) [inline network-protection/network-protection-api/src/main/java/com/duckduckgo/networkprotection/api/NetworkProtectionAccessState.kt:40]: this change needs to be documented and be known to others. follow the API change process

**Review by malmstein — APPROVED** [review]
Works as expected, check with @aitorvs  and @karlenDimla  for changes related to NetP API.

**Review by malmstein — COMMENTED** [review]
As discussed offline, the only thing I’d look into changing is to move all the UI States into the impl module because they don’t need to be exposed.

**Comment by mikescamell** [comment]
> Works as expected, check with @aitorvs and @karlenDimla for changes related to NetP API.

Please hold off reviewing until I've made some changes, I think we can move the new NetworkProtectionAccessState API to the impl module as no other module is using it.

**Comment by aitorvs** [comment]
> > Works as expected, check with @aitorvs and @karlenDimla for changes related to NetP API.
> 
> Please hold off checking until I've made some changes, I think we can move the new NetworkProtectionAccessState API to the impl module as no other module is using it.

I will review the changes. 
We should not need to change the VPN public API to revamp settings. 
And I've seen some odd changes, for instance we're moving subscription state to the netp public API, that should not be.

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/subscription/NetpSubscriptionManager.kt [hunk 2]
```diff
@@ -70,9 +74,21 @@ class RealNetpSubscriptionManager @Inject constructor(
 
     private fun hasValidEntitlementFlow(): Flow<Boolean> = subscriptions.getEntitlementStatus().map { it.contains(NetP) }
 
-    private suspend fun getVpnStatusInternal(hasValidEntitlement: Boolean): VpnStatus {
+    private suspend fun getVpnStatusInternal(hasValidEntitlement: Boolean): VpnStatus = if (newSettingsFeature.self().isEnabled()) {
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/subscription/NetpSubscriptionManager.kt:77]: Please add brackets to the method. I know kotlin allows for this but it harms readability.

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/subscription/settings/NetworkProtectionSettingsState.kt [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+import kotlinx.coroutines.flow.Flow
+
+interface NetworkProtectionSettingsState {
+
+    /**
+     * Returns a flow of the visibility states of NetP
+     * The caller DOES NOT need to specify the dispatcher when calling this method
+     */
+    suspend fun getStateFlow(): Flow<NetPSettingsState>
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/subscription/settings/NetworkProtectionSettingsState.kt:27]: nit: rename to `getSettingsStateFlow()`?

File: network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/subscription/settings/NetworkProtectionSettingsStateImpl.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    private val networkProtectionState: NetworkProtectionState,
+    private val netpSubscriptionManager: NetpSubscriptionManager,
+) : NetworkProtectionSettingsState {
+
+    override suspend fun getStateFlow(): Flow<NetPSettingsState> =
+        netpSubscriptionManager.vpnStatus().map { status ->
+            if (!status.isActive()) {
+                // if entitlement check succeeded and not an active subscription then reset state
+                handleRevokedVPNState()
+            }
```
Comment (aitorvs) [inline network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/subscription/settings/NetworkProtectionSettingsStateImpl.kt:55]: Should this logic be here?
If I understand correctly, a user can have an inactive subscription but continue using the VPN as long as they don't open settings? That seems odd/wrong?

Comment (mikescamell) [reply network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/subscription/settings/NetworkProtectionSettingsStateImpl.kt:55]: I'm taking the logic that already exists in `NetworkProtectionAccessStateImpl`  and adding it here.

```
   override suspend fun getStateFlow(): Flow<NetPAccessState> = withContext(dispatcherProvider.io()) {
        netpSubscriptionManager.vpnStatus().map { status ->
            if (!status.isActive()) {
                // if entitlement check succeeded and no entitlement, reset state and hide access.
                handleRevokedVPNState()
                Locked
            } else {
                UnLocked
            }
        }
    }
```

This is only ever called by the current NetP Settings item: `ProSettingNetPViewModel` (`LegacyProSettingNetPViewModel` in this PR)

Once the new Settings is rolled out, `NetworkProtectionAccessStateImpl.getStateFlow` will never be called, so I wanted to ensure the behaviour stays the same when that eventually happens.

Comment (mikescamell) [reply network-protection/network-protection-impl/src/main/java/com/duckduckgo/networkprotection/impl/subscription/settings/NetworkProtectionSettingsStateImpl.kt:55]: As discussed on Zoom, this is existing behaviour in the Legacy `ProSettingNetPViewModel` and therefore I'm going to leave it as is so as not to affect existing behaviour.

File: network-protection/network-protection-api/src/main/java/com/duckduckgo/networkprotection/api/NetworkProtectionAccessState.kt (hunk 1 above)
Comment (mikescamell) [reply network-protection/network-protection-api/src/main/java/com/duckduckgo/networkprotection/api/NetworkProtectionAccessState.kt]: I've removed this and moved the state internally. See: [`82174dd` (#5394)](https://github.com/duckduckgo/Android/pull/5394/commits/82174dd47ee45383775988a0456a2a34aecb16d2)

**Pull request merged**